feat: HID Usages + Keys Standardisation

innovaker opened the pull request:
- Adds HID Usages for ZMK (#217).
- Standardizes key aliases/codes (#21 - still open for RFC).
- Enhances HID reports to take advantage of the new codes (#234).

### Issues
- Closes #217.
- Closes #21.
- Closes #234.
- Supersedes and closes #246.
- Supersedes and closes #158.
- Supersedes and closes #161.

### Testing
- End-to-end tests covering:
  - all key aliases
  - both the keyboard (keypad) and consumer HID reports
- ZMK on Nordic's nRF52840 Dongle.
  - `kscan-mock` running on dongle systematically pressed and released each key.
- Node scripts on Raspberry Pi (Raspbian Buster)
   - Scripts running on the Pi parsed events from `/dev/hidrawX` and `/dev/input/...` devices and compared incoming data against expected key usage codes.
   - Scripts not included in PR because end-to-end framework isn't generic enough and some of the code is my own.
- Test logs:
  - [e2e-2020-10-04T12-14-37-usb.log](https://github.com/zmkfirmware/zmk/files/5323138/e2e-2020-10-04T12-14-37-usb.log)
  - [e2e-2020-10-04T12-14-37-ble.log](https://github.com/zmkfirmware/zmk/files/5323139/e2e-2020-10-04T12-14-37-ble.log)
- Once HID reports were enhanced, usage codes themselves never caused any errors.
- Untested usage codes should also be viable for the future as everything was developed systematically via scripting.

innovaker commented:
Regarding HID usage codes, the keen observers will notice that I included all the usages - even the ones we don't utilize yet.

This is because:
a) it keeps doors open for the future (I personally plan on using some of them myself).
b) it wasn't much extra work.

I can remove them and refresh the PR if people really don't want them included for whatever reason.

innovaker commented:
I suggest we give this PR at least a few days to settle because there may be some developers that missed the RFC for #21.  It would encourage multi-platform testing too.

innovaker commented:
@tominabox1 has encountered unintended consequences of 2da9078ac43f7e38205b7a231e54df1520862ac7 when users have defined their own behaviour instances: https://discordapp.com/channels/719497620560543766/719544230497878116/762294497081163806
i.e. in their keymaps

I think it's still worth deprecating `USAGE_KEYPAD` and `USAGE_CONSUMER` for their replacements (see `hid_usage.h` to see what I mean).  But should we:
- ask our more advanced users to update their keymaps on merge of this issue?
- provides legacy fallback defines - i.e. "symlinks" in `hid_usage_legacy.h` - for the short term?

Preference on this @petejohanson please?

For anyone trying to test, in the meantime the replacements are:
`USAGE_KEYPAD`: `HID_USAGE_KEY`
`USAGE_CONSUMER`: `HID_USAGE_CONSUMER`

innovaker commented:
Follow-up to my last comment ... it's mainly a concern for anyone who has merged https://github.com/zmkfirmware/zmk/pull/153 into their local as it uses:
- `USAGE_KEYPAD`: 20 times
- `USAGE_CONSUMER`: 6 times

innovaker commented:
I've force-pushed a new version that adds legacy values for `USAGE_KEYPAD` and `USAGE_CONSUMER`.  @tominabox1, could you give it a try please (please remember to strip out the previous pull)?

innovaker commented:
@tominabox1 has gracefully retested.  There are a few merge conflicts with #153.  So if you're experimenting with @okke-formsma's modifiers branch and are looking to test it alongside this PR, then you'll need to resolve those merge conflicts.  The force push earlier reintroduced `USAGE_KEYPAD` and `USAGE_CONSUMER` as legacy values for this purpose.

innovaker commented:
Most recent test logs from the re-push:
- [e2e-2020-10-04T15-43-02-ble.log](https://github.com/zmkfirmware/zmk/files/5323556/e2e-2020-10-04T15-43-02-ble.log)
- [e2e-2020-10-04T15-43-02-usb.log](https://github.com/zmkfirmware/zmk/files/5323557/e2e-2020-10-04T15-43-02-usb.log)

innovaker commented:
As I develop the associated documentation, I'm starting to notice small areas for improvement.  So I'll probably re-push some changes after pushing the first draft of #218 (coming soon).

petejohanson reviewed (commented):
This really looks solid. Looking forward to the updated version for final review.

innovaker commented:
### Updates
- Updated usages to HID Usage Tables 1.21
  - See https://usb.org/sites/default/files/hut1_21.pdf.
- Deprecated codes found in bug #246.
- Added `K_` convention.
  - See https://github.com/zmkfirmware/zmk/issues/21#issuecomment-706690476.
- Added extra codes discovered by scanning `0x000` to `0xFFF` through `evtest` (both pages).
- Minor improvements to various names.
- Integrated RFC improvements from @petejohanson and @innovaker.
  - See discussions in https://docs.google.com/spreadsheets/d/1s6URDIL7pqcMiUvY8s0esWvLIA1dKNmHqp6jEUtrqTw/.
- Refactored legacy names into inline deprecated names with inline warnings.
- Replaced deprecated names in code, tests and documentation.
- Refactored names in code and tests with fully-qualified names.
- Reformatted example keymaps in documentation to better fit new key names.
- Reordered commits to be more logical and `git blame` friendly.
- Refactored commit messages for improved readability.

### Testing
- app
  - Same as before.
  - logs:
    - [e2e-2020-10-28T11-49-58-ble.log](https://github.com/zmkfirmware/zmk/files/5454672/e2e-2020-10-28T11-49-58-ble.log)
    - [e2e-2020-10-28T11-49-58-usb.log](https://github.com/zmkfirmware/zmk/files/5454673/e2e-2020-10-28T11-49-58-usb.log)
- docs: 
  - `eslint . --ext js,jsx,md,mdx`
  - `npx prettier --write .`

### Reviews / Feedback
- Please pull and test it, your existing keymap *should* still work!!!
- Any feedback or suggestions welcomed!
- There was a lot of data involved, can you spot any oversights?
- For reviewing the key names, please review in conjunction with PR #308 (it will make it easier!).

tominabox1 commented:
I've tried mapping the media keys (cp and kp varieties) and can't seem to get them to work on my laptop anymore. I remember we tested them a few weeks ago and they worked so I'm not sure what may have changed.

innovaker commented:
Thanks @tominabox1.  You might also try the various other[ media key codes](https://5f99f3c599fd970007c1451a--zmk.netlify.app/docs/codes/media).  Before I settled on the names, I did cursory testing on applications within Android, Linux and Windows and discovered the support to be inconsistent and occasionally not what you might expect.  More on that here if you're curious: https://github.com/zmkfirmware/zmk/issues/21#issuecomment-702182102.

There could also be difference between what the OS supports and which applications support them.  I touched on that in this [message](https://discord.com/channels/719497620560543766/719544230497878116/771342122834067496) on Discord.

adamcstephens commented:
Volume works for me across Linux and Mac. Will need to get some time trying some of the other media keys. Otherwise, I'm using this branch as my daily driver and will report any other issues.

petejohanson commented on app/include/zmk/hid.h:
```diff
@@ -85,7 +86,7 @@ static const u8_t zmk_hid_report_desc[] = {
     /* REPORT_SIZE (8) */
     HID_GI_REPORT_SIZE,
-    0x02,
+    0x01,
```
In doing some further testing on this, when using my MDK kit over USB, I do *not* get key events with this change and the change to the `ZMK_HID_MAX_KEYCODE` value. My *guess* is that this is related to the USB data being too large after this change, and causing problems for the onboard USB HUB, perhaps?

@innovaker Thoughts on reverting just these changes, in favor of switching to a `6KRO` (and future `#KRO`) type HID report/descriptor for this in a future PR?

Doing so would of course limit the actual *use* of some of these new HID keycodes defined here, but allow us to move forward with other changes while we work on iterating on the actual transmitted HID report encoding/format.

innovaker replied:
Casual observers, this was discussed throughout the day on Discord.  We replaced a core part of the system in an effort to fix this issue.  More delays follow below.

innovaker commented:
Heads up to anyone currently testing/exploring/using this PR!

Further to Pete's earlier feedback, today we replaced a core part of the system on which this PR relies.  It's currently undergoing testing as PR #313.

Once that's ready and merged, I'll then re-push this PR rebased on top `main`'s new `HEAD`, and we'll kindly ask you to try the refreshed version.  It'll likely be out sometime tomorrow or the weekend.

Thanks for testing it, we appreciate the feedback!

innovaker commented:
### Changes
- Rebased onto #313 and #269.
- Refactored HID pages into `hid_usage_pages.h`.
- Replaced Zephyr usages with ZMK usages in HID reports.
- Minor comments tweaks.

### Testing
- As before.
  - [e2e-2020-10-31T16-05-57-ble.log](https://github.com/zmkfirmware/zmk/files/5469394/e2e-2020-10-31T16-05-57-ble.log)
  - [e2e-2020-10-31T16-05-57-usb.log](https://github.com/zmkfirmware/zmk/files/5469395/e2e-2020-10-31T16-05-57-usb.log)

innovaker commented:
@tominabox1, @adamcstephens, @okke-formsma, anyone else testing - should be good to go! :)

okke-formsma commented:
lgtm

innovaker commented:
Having now laid the groundwork for shifted keys (a future PR), a minor quandary presented itself for *this* issue that we didn't previously consider.

What should we call `'` and `"`?
 - `APOSTROPHE`/`APOS` and `QUOTE`?
 - `SINGLE_QUOTE`/`SQT` and `DOUBLE_QUOTE`/`DQT`?
 - `QUOTE`/`QUOTE` and `DOUBLE_QUOTE`/`DQUOTE`?
 - `QUOTE` and `QUOTE2`?
 - Should we use a mix?
 - Some other idea I haven't considered?

@petejohanson, any thoughts please?

tominabox1 commented:
I would prefer to see us use something similar to (if not the same as) QMK because I suspect it will be most familiar to new zmk users. They use "QUOTE" - so maybe we use "QUOTES" that way we cover "single" and "double" quotes.

innovaker commented:
From what I've observed, it seems to be a toss up between:
 - `QUOTE`/`QT` and `QUOTES`/`QTS`
 - `SINGLE_QUOTE`/`SQT` and `DOUBLE_QUOTE(S)`/`DQT`

but also including `APOSTROPHE`/`APOS` for the Brits regardless.

@petejohanson, you favor the latter I believe?  If so, preference on plural vs. singular for the double please?

innovaker commented:
With the shifted keys work complete (that's the second part of #86, *not* #153), if we are able to finalise their names soon then I'm considering just rolling their *names* into this PR - but *without* the modifier `LS()` wrappers until #153 has completed its review process.  So effectively they'd just an alias of their non-shifted key.

Why?  Because of #246.  This PR currently perpetuates the problems of #246 because they were incorrectly mapped in the original version.  We kept them in until now because we didn't have an alternative.  Now we do, so I'd rather get them in.  The legwork is done, we just need to agree on the naming scheme.

innovaker commented:
### Changes
 - Added *shifted key name* candidates.
   - This effectively "corrects" the identities for the remainder of #246 (except `CURU` which is being removed).
   - Once modifiers #153 has completed its review cycle, the `LS(...)` wrappers will be added to the shifted key names.
   - The ongoing RFC for the shifted key names (**green** cells) can be found here:
     https://docs.google.com/spreadsheets/d/1s6URDIL7pqcMiUvY8s0esWvLIA1dKNmHqp6jEUtrqTw/

### Testing
 - As before.
   - [e2e-2020-11-01T14-01-31-ble.log](https://github.com/zmkfirmware/zmk/files/5470991/e2e-2020-11-01T14-01-31-ble.log)
   - [e2e-2020-11-01T14-01-31-usb.log](https://github.com/zmkfirmware/zmk/files/5470992/e2e-2020-11-01T14-01-31-usb.log)


### Notes
 - This could be the last push unless (either):
   - The shifted keys RFC introduces minor changes.
   - I come across any problems (unlikely) during my final self-review.
 - Testers, your efforts have been much appreciated!  You're welcome to retest this version if you'd like.  There shouldn't be any noticeable differences ... unless I screwed up 👍

joelspadin commented:
My main concern with using `QUOTE` and `QUOTES` is how similar they are. It would be easy to make a typo and get the wrong one, and I'm not sure it's very clear what the difference is, especially given that on a standard keyboard the legend on the key is two different quote symbols and thus "quotes" in the sense of "multiple kinds of quote" instead of "double quote".

My vote is for making single vs. double explicit. Alternatively, something like `QUOTE` for `'` and `DOUBLE_QUOTE` for `"` would also avoid that confusion, but I don't think that's as consistent.

innovaker commented on app/include/zmk/hid_usage_pages.h:
```diff
@@ -0,0 +1,47 @@
+#define USAGE_CONSUMER (0x0C) // WARNING: DEPRECATED (DO NOT USE)
+
+#define HID_USAGE_GD (0x01)             // Generic Desktop
+#define HID_USAGE_SIM (0x02)            // Simulation Controls
+#define HID_USAGE_VR (0x03)             // VR Controls
+#define HID_USAGE_SPORT (0x04)          // Sport Controls
+#define HID_USAGE_GAME (0x05)           // Game Controls
+#define HID_USAGE_GDV (0x06)            // Generic Device Controls
+#define HID_USAGE_KEY (0x07)            // Keyboard/Keypad
+#define HID_USAGE_LED (0x08)            // LED
+#define HID_USAGE_TELEPHONY (0x0B)      // Telephony Device
+#define HID_USAGE_CONSUMER (0x0C)       // Consumer
+#define HID_USAGE_DIGITIZERS (0x0D)     // Digitizers
+#define HID_USAGE_HAPTICS (0x0E)        // Haptics
+#define HID_USAGE_PID (0x0F)            // PID
+#define HID_USAGE_EHT (0x12)            // Eye and Head Trackers
+#define HID_USAGE_AUXDISP (0x14)        // Auxiliary Display
+#define HID_USAGE_SENSORS (0x20)        // Sensors
+#define HID_USAGE_MEDICAL (0x40)        // Medical Instrument
+#define HID_USAGE_BRAILLE (0x41)        // Braille Display
+#define HID_USAGE_LIGHT (0x59)          // Lighting And Illumination
+#define HID_USAGE_MONITOR (0x80)        // USB Monitor
+#define HID_USAGE_MONITOR_VALUES (0x81) // Monitor Enumerated Values
+#define HID_USAGE_MONITOR_VESA (0x82)   // VESA Virtual Control
+#define HID_USAGE_POWER (0x84)          // Power
+#define HID_USAGE_POS_BARCODE (0x8C)    // Bar Code Scanner
+#define HID_USAGE_POS_SCALE (0x8D)      // Scale
+#define HID_USAGE_POS_MSR (0x8E)        // Magnetic Stripe Reading (MSR) Devices
+#define HID_USAGE_POS_RESV (0x8F)       // Reserved Point of Sale
+#define HID_USAGE_CAMERA (0x90)         // Camera Control
+#define HID_USAGE_ARCADE (0x91)         // Arcade
+#define HID_USAGE_GAMING (0x92)         // Gaming Device
+#define HID_USAGE_FIDO (0xF1D0)         // FIDO Alliance
```
I tossed up between a few conventions for the usage pages:
 - `HID_USAGE_XXXXX`
 - `HID_USAGE_PAGE_XXXXX`
 - `HID_USAGE_XXXXX_PAGE`

I went for the first because the form facilitated natural namespaces for the HID usage definitions (see `hid_usage.h`).

But having now put the pages into the HID reports, they can be a bit confusing.   So I'm honestly not sure which is best from a readability perspective. 

Any thoughts from anyone please?

petejohanson replied:
I think I favor the way you have it here. I think it's plenty clear.

innovaker commented:
@tominabox1 , @adamcstephens, @okke-formsma.  Depending on what comes out of tonight (US time), it's looking like final testing will be tomorrow, probably towards the evening as people will be working.  Any changes between now and then should be minor though, unless something goes wrong.

There's no expectations of you, but I thought I'd give you the heads up nonetheless.  Thanks for all your support so far!

tominabox1 commented:
Testing a number of codes: 
International codes don't seem to do anything on my MacOS Catalina system (may not know how they work). 
`C_VOL_UP` works but `K_VOL_UP` does not (nor does `K_VOL_UP2`)
Interestingly, `C_MUTE` and `K_MUTE` works, but `K_MUTE2` does not
Eject keys don't trigger any behavior on my system
`CLEAR` and `PRIOR` don't seem to do anything
`C_ALT_AUDIO_INC` `C_PLAY` Don't do anything
 `C_PP` Works
`C_REC` doesn't trigger anything
 `K_PP` `C_PAUSE` `C_STOP` `K_STOP2` `K_STOP3` doesn't do anything
`C_STOP_EJECT`doesn't do anything
`C_NEXT` works
`K_NEXT` doesn't work
`C_PREV` works
`K_PREV` doesn't work
`C_FF` `C_RW` Works - same behavior as next/prev
`C_SHUFFLE` doesn't work
Also tested editing `cp` codes (cut/copy/paste) and did't see any behaviors on the host system.

innovaker commented:
Thanks a lot @tominabox1!  That's roughly what I'd expect from mac OS given what I've picked up in passing along this journey.

innovaker commented:
### Changes
- 3x shifted key names updated after RFC feedback.

### Testing
- As before.
  - [e2e-2020-11-02T09-41-02-ble.log](https://github.com/zmkfirmware/zmk/files/5474006/e2e-2020-11-02T09-41-02-ble.log)
  - [e2e-2020-11-02T09-41-02-usb.log](https://github.com/zmkfirmware/zmk/files/5474007/e2e-2020-11-02T09-41-02-usb.log)

innovaker commented:
Please don't merge this issue until I sign it off.

petejohanson commented on app/include/dt-bindings/zmk/keys.h:
```diff
@@ -6,143 +6,1387 @@
+#define QUOT (SINGLE_QUOTE) // WARNING: DEPRECATED (DO NOT USE)
+
+/* Keyboard " (Quote) */
+#define DOUBLE_QUOTES (HID_USAGE_KEY_KEYBOARD_APOSTROPHE_AND_QUOTE)
```
With all of these added "shifted codes", would the replacing these with the shifted varients once that's merged. e.g.:

```suggestion
#define DOUBLE_QUOTES (LS(HID_USAGE_KEY_KEYBOARD_APOSTROPHE_AND_QUOTE))
```

?

That's really my only outstanding item I would want clarity on. Otherwise this looks great.

CrossR replied:
The rebased branch with modifiers and this PR has these keycodes with the shifted variant: https://github.com/innovaker/zmk/tree/hid-keys-modifiers-rebase. I.e. 

https://github.com/innovaker/zmk/blob/6e52af9870ddb88e05a948e4d9160911e6543e9f/app/include/dt-bindings/zmk/keys.h#L278-L280

So I expect that is the plan, though it was emphasised that branch is a "dry run" so there may be some other tiny bits of work around integrating them I guess.

(Caveat aside, I've been using that  branch (hid-keys-modifiers-rebase) all of today with no issues except a few expected collisions between names and my own defines when doing the build, which is expected. Made my keymap a lot shorter, I only need defines for UK keycodes now, as the names don't make much sense there).

petejohanson replied:
@CrossR Indeed. Good catch. Thanks.

petejohanson reviewed (commented):
Just one question on the shifted codes.

petejohanson reviewed (approved):
@innovaker This looks good to me. Ready to merge when you are.

innovaker commented:
Thanks for looking through it.  I'll be doing the rebase today.

innovaker commented:
### Changes
 - Rebased onto `main`'s `HEAD`.
 - Aligned modifier PR #153 to this PR's changes.
 - Added shifting wrappers `LS()` to shifted key definitions.

### Testing
- As before (same checklist), but also tested with the deprecated codes completely removed.
- end-2-end tests (including deprecated codes):
  - [e2e-2020-11-03T14-03-40-ble.log](https://github.com/zmkfirmware/zmk/files/5482146/e2e-2020-11-03T14-03-40-ble.log)
  - [e2e-2020-11-03T14-03-40-usb.log](https://github.com/zmkfirmware/zmk/files/5482147/e2e-2020-11-03T14-03-40-usb.log)
- end-2-end tests (without deprecated codes):
  - [e2e-2020-11-03T13-54-35-ble.log](https://github.com/zmkfirmware/zmk/files/5482142/e2e-2020-11-03T13-54-35-ble.log)
  - [e2e-2020-11-03T13-54-35-usb.log](https://github.com/zmkfirmware/zmk/files/5482143/e2e-2020-11-03T13-54-35-usb.log)

### Notes
 - This is the release candidate!
  So, it will not change unless either:
   - Testing uncovers problems (as it did last Friday).
   - I suddenly spot something important that I missed (my checklist was huge, so it's unlikely).
 - The rebase had some conflicts and this is a big PR, so I recommend:
   - Give it a day to "settle" before merging.
     - This should also give me enough time to prepare the documentation for simultaneous release which be a big bonus.
   - Testers *sign off* before it's merged.
 - I appreciate everyone is anxious to get this out.  I personally can't wait to dust my hands of it.  But *please* be patient!  We just need one more day for stability testing.

### Testers
 - Please *leave a comment* if you're intending to test this release candidate.  We need visibility of what's going on.

petejohanson commented on app/include/dt-bindings/zmk/keys.h:
```diff
@@ -5,142 +5,1385 @@
  */
 #pragma once
 
+#include <zmk/hid_usage.h>
```
Actually just noticed this one *organiational* issue. In general, I've been keeping *all* headers that might be included into keymaps, and are by requirement *only* preprocessor directives, in the `include/dt-bindings/` directory so there is a clear delineation of "headers used in keymaps". Those headers of course can/will be used by the code, but it's clear they are used in keymaps this way.

@innovaker Any problems moving the couple files into those directories? Shouldn't require re-testing, beyond just verifying builds, and would help keep that distinction intact.

innovaker replied:
To confirm:
- `zmk/hid_usage.h` to `dt-bindings/zmk/hid_usage.h`
- `zmk/hid_usage_pages.h` to `dt-bindings/zmk/hid_usage_pages.h`
?

petejohanson replied:
That's the only two I can see, yeah.

innovaker replied:
Any preference on the `#include` orders for those?

innovaker replied:
nevermind on my last

innovaker replied:
Consider it done.

innovaker commented:
Hold my beer, @petejohanson has requested an organisational change.  Will re-push in 2 hours.

innovaker commented:
Release Candidate ... 2?

### Changes
 - Moved `hid_usage.h` and `hid_usage_page.h` from `include/zmk` into `include/dt-bindings/zmk` as per @petejohanson's [request](https://github.com/zmkfirmware/zmk/pull/236#discussion_r516802721).

### Testing
 - As before.
 - end-to-end tests (including deprecated codes):
   - [e2e-2020-11-03T18-22-05-ble.log](https://github.com/zmkfirmware/zmk/files/5483252/e2e-2020-11-03T18-22-05-ble.log)
   - [e2e-2020-11-03T18-22-05-usb.log](https://github.com/zmkfirmware/zmk/files/5483253/e2e-2020-11-03T18-22-05-usb.log)

### Testers
 - Please *leave a comment* if you're intending to test this release.  Thanks!

innovaker commented:
Release Candidate 3 ... the final?  Cause I'm off to bed.

### Changes
 - Rebased onto #325, because it's useful and 10 is a lucky number (this will be the last rebase!!!)

### Testing
 - As before.
 - end-to-end tests (including deprecated codes):
   - [e2e-2020-11-03T20-17-12-ble.log](https://github.com/zmkfirmware/zmk/files/5483848/e2e-2020-11-03T20-17-12-ble.log)
   - [e2e-2020-11-03T20-17-12-usb.log](https://github.com/zmkfirmware/zmk/files/5483849/e2e-2020-11-03T20-17-12-usb.log)

### Testers
 - Please *leave a comment* if you're intending to test this release.  Thanks!

innovaker commented:
I believe @tominabox1 is doing some testing.

adamcstephens commented:
This works for me. Thanks!

petejohanson commented:
@tominabox1 has noted on Discord it was working fine for him.

chenkevinh commented:
Seems to be working fine for me on Windows so far.

petejohanson commented:
Also got some testing verification from @Nicell w/ his nice60, no reported problems. Merging this!
